server/src/internalClusterTest/java/org/opensearch/index/SegmentReplicationPressureIT.java

```diff
@@ -0,0 +1,300 @@
+            });
+            indexingThread.start();
+            indexingThread.join();
+            latch.await();
+            // try and index again while we are stale.
+            assertBusy(() -> { assertFailedRequests(executeBulkRequest(nodes, randomIntBetween(1, 200))); });
+        }
+        refresh(INDEX_NAME);
+        // wait for the replicas to catch up after block is released.
+        waitForSearchableDocs(totalDocs, replicaNodes.toArray(new String[] {}));
+
+        // index another doc showing there is no pressure enforced.
+        executeBulkRequest(nodes, totalDocs);
+        waitForSearchableDocs(totalDocs * 2L, replicaNodes.toArray(new String[] {}));
+        verifyStoreContent();
+    }
+
+    private BulkResponse executeBulkRequest(List<String> nodes, int docsPerBatch) {
+        final BulkRequest bulkRequest = new BulkRequest();
+        for (int j = 0; j < docsPerBatch; ++j) {
+            IndexRequest request = new IndexRequest(INDEX_NAME).id(UUIDs.base64UUID())
+                .source(Collections.singletonMap("key", randomAlphaOfLength(50)));
+            bulkRequest.add(request);
+        }
+        final BulkResponse bulkItemResponses = client(randomFrom(nodes)).bulk(bulkRequest).actionGet();
+        refresh(INDEX_NAME);
+        return bulkItemResponses;
+    }
+
+    /**
+     * Index and Refresh in batches to force checkpoints behind.
+     * Asserts that there are no stale replicas according to the primary until cp count is reached.
+     */
+    private int indexUntilCheckpointCount() {
+        int total = 0;
+        for (int i = 0; i < MAX_CHECKPOINTS_BEHIND; i++) {
+            final int numDocs = randomIntBetween(1, 100);
+            for (int j = 0; j < numDocs; ++j) {
+                indexDoc();
+            }
+            total += numDocs;
+            refresh(INDEX_NAME);
+        }
+        return total;
+    }
+
+    private void assertFailedRequests(BulkResponse response) {
+        assertTrue(Arrays.stream(response.getItems()).allMatch(BulkItemResponse::isFailed));
+        assertTrue(
+            Arrays.stream(response.getItems())
+                .map(BulkItemResponse::getFailure)
+                .allMatch((failure) -> failure.getStatus() == RestStatus.TOO_MANY_REQUESTS)
+        );
+    }
+
+    private void indexDoc() {
+        client().prepareIndex(INDEX_NAME).setId(UUIDs.base64UUID()).setSource("{}", "{}").get();
+    }
+
+    private void assertEqualSegmentInfosVersion(List<String> replicaNames, IndexShard primaryShard) {
+        for (String replicaName : replicaNames) {
+            final IndexShard replicaShard = getIndexShard(replicaName, INDEX_NAME);
+            assertEquals(
+                primaryShard.getLatestReplicationCheckpoint().getSegmentInfosVersion(),
+                replicaShard.getLatestReplicationCheckpoint().getSegmentInfosVersion()
+            );
+        }
+    }
+
+    private Releasable blockReplication(List<String> nodes, CountDownLatch latch) {
```

Comment:
nit: This is useful and common utility which is used in different forms in different SegmentReplication* test classes. It can be moved to parent class.

Reply:
moved this to base class.